### PR TITLE
Added CODING_CONVENTIONS_C++.md

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

- Copied content from https://github.com/RIOT-OS/RIOT/wiki/Coding-Conventions-Cplusplus/8d2a6c8a0c744dcca6357e1a40d7bfeadf3ec31c to `CODING_CONVETIONS_C++.md`
- Fixed some typos and white space errors
- Improved some formatting of the markdown
    ==> Except for the fixed typos, the output when running a markdown parser  should be identical to the code in the wiki

### Testing procedure

- Run a markdown parser once with the newly added `CODING_CONVENTIONS_C++.md` file as input and once with the code of the wiki page as input
- Use `diff` to compare the two outputs
    ==> The difference should only be fixed typos.

### Issues/PRs references

Follow up of https://github.com/RIOT-OS/RIOT/pull/12572

### CODE REVIEW

**Comment by maribu** [comment]
I think as no maintainer complained about moving the C coding convention and this is basically the same, requesting a review from all maintainers would be almost SPAM.

**Comment by miri64** [comment]
Could you at least split up the moving and spell-fix+cleanup part into two seperate commits? This would make it easier to verify, that the wiki entry was moved unchanged.

File: CODING_CONVENTIONS_C++.md [hunk 1]
```diff
@@ -0,0 +1,369 @@
+You should check out the [C Conventions](CODING_CONVENTIONS.md) as some section
+still apply (Documentation, Git, Travis).
+
+When contributing source code, please adhere to the following coding style,
+which is loosely based on the [Google C++ Style Guide] and the coding
+ conventions used by the C++ Standard Library. Based on the [CAF coding style].
+
+[Google C++ Style Guide]: (http://google-styleguide.googlecode.com/svn/trunk/cppguide.xml)
```
Comment (gschorcht) [inline CODING_CONVENTIONS_C++.md]: The link doesn't work. Should be probably https://google.github.io/styleguide/cppguide.html

File: CODING_CONVENTIONS_C++.md [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+[CAF coding style]: (https://github.com/actor-framework/actor-framework/blob/develop/CONTRIBUTING.md)
```
Comment (gschorcht) [inline CODING_CONVENTIONS_C++.md]: This link doesn't work too, should be https://github.com/actor-framework/actor-framework/blob/master/CONTRIBUTING.md.

File: CODING_CONVENTIONS_C++.md [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+  For example, the header file for `riot::example::my_class` of module
+  `my_module` is located at `path/to/my_module/riot/example/my_class.hpp`
+  and the source file at `path/to/my_module/my_class.cpp`.
+
+- All header files should use `#define` guards to prevent multiple inclusion.
+  The symbol name is `<RELATIVE>_<PATH>_<TO>_<FILE>_HPP`.
+
+- Do not `#include` when a forward declaration suffices.
+
+- Each library component must provide a `fwd.hpp` header providing forward
```
Comment (gschorcht) [inline CODING_CONVENTIONS_C++.md]: This is directly taken from the CAF coding style but makes IMHO not really sense. The standard way to include all headers of a library `xyz` with only one include statement is to define a header file `xyz/xyz.hpp`. This rule should be revised.

File: CODING_CONVENTIONS_C++.md [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+  declarations for all types used in the user API.
+
+- Each library component must provide an `all.hpp` header that contains the
```
Comment (gschorcht) [inline CODING_CONVENTIONS_C++.md]: Forward declarations for a library are usually done in `xyz/xyz.hpp` too, see above.

**Comment by gschorcht** [comment]
IMHO, the style guide should be revised. From my point of view, the official C++ standard and coding examples at https://en.cppreference.com/w/cpp/language/ should be used as reference instead of vendor specific or project specific style guides. But this is out of the scope of this PR.

**Comment by maribu** [comment]
I fixed the broken links and directly squashed

 I agree with the content changes @gschorcht suggested. I don't know why we should build upon CAF coding conventions at all. In the C coding conventions we similarly build upon the Linux coding convention, but this can easily be justified: The Linux kernel coding conventions are targeting a similar audience, are well established, and C programmers can be expected to be at least aware of them.

CAF on the other hand seems to be a minor player, is not as obviously related to RIOT (in terms of similar goals) as Linux is, and a minor player cannot be well established.

However: I also agree that this should be a separate PR. Also: I'm not a C++ guy, so I'll better let people with more qualification on that topic do the discussion.

**Comment by miri64** [comment]
On the commit message of f22174992571e66ab180888399fda158a1e44b23: rather link to the [latest revision](https://github.com/RIOT-OS/RIOT/wiki/Coding-Conventions-Cplusplus/8d2a6c8a0c744dcca6357e1a40d7bfeadf3ec31c) than the edit link, so we have a reference for the future ;-).

**Comment by miri64** [comment]
(same goes for OP)

**Comment by maribu** [comment]
@gschorcht: I think I addressed your concerns (with the exception of the content changes, that you proposed to handle in a separate PR).

**Comment by gschorcht** [comment]
> @gschorcht: I think I addressed your concerns (with the exception of the content changes, that you proposed to handle in a separate PR).

Yeap. @miri64 is right, we should link to a certain commit, sorry my fault.

**Comment by miri64** [comment]
> CAF on the other hand seems to be a minor player, is not as obviously related to RIOT (in terms of similar goals) as Linux is, and a minor player cannot be well established.

I think the relation is: the people who did most of the C++ support for RIOT also developed CAF.

**Comment by miri64** [comment]
Ping @gschorcht?

File: CODING_CONVENTIONS_C++.md [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+- All names except macros and template parameters should be
+  lower case and delimited by underscores.
+
+- Template parameter names should be written in CamelCase.
+
+- Types and variables should be nouns, while functions performing an action
+  should be "command" verbs. Classes used to implement metaprogramming
+  functions also should use verbs, e.g., `remove_const`.
+
+- Private and protected member variables use the suffix `_` while getter *and*
```
Comment (gschorcht) [inline CODING_CONVENTIONS_C++.md]: Please remove the trailing whitespace here to make Travis happy.

Comment (maribu) [reply CODING_CONVENTIONS_C++.md]: Done and squashed

**Comment by maribu** [comment]
(I also found a line that was longer than 80 chars I didn't break before. I fixed that, too. See the "force-pushed" for the diff.)

**Pull request merged**